src/os/src/os_log.c

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright(c) 2006 to 2018 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <assert.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "os/os.h"
+
+#define MAX_TIMESTAMP_LEN (10 + 1 + 6)
+#define MAX_TID_LEN 10
+#define MAX_HDR_LEN (MAX_TIMESTAMP_LEN + 1 + MAX_TID_LEN + 2)
+
+#define BUF_OFFSET MAX_HDR_LEN
+
+typedef struct {
+  char buf[2048];
+  size_t bufsz;
+  size_t pos;
+} log_buffer_t;
+
+typedef struct {
+  dds_log_write_fn_t func;
+  void *ptr;
+} log_sink_t;
+
+static os_threadLocal log_buffer_t log_buffer;
+
+static os_once_t lock_inited = OS_ONCE_T_STATIC_INIT;
+static os_rwlock lock;
+
+static uint32_t log_mask = DDS_LC_ERROR | DDS_LC_WARNING;
+
+static FILE *log_file;
+static FILE *trace_file;
+
+static log_sink_t log_sink = { .func = NULL, .ptr = NULL };
+static log_sink_t trace_sink = { .func = NULL, .ptr = NULL };
+
+uint32_t *const dds_log_mask = &log_mask;
+
+static size_t print_header(log_buffer_t *lb)
+{
+    int cnt;
+    char *tid, buf[MAX_TID_LEN] = { 0 };
+    char hdr[MAX_HDR_LEN + 1];
+    static const char fmt[] = "%d.%06d/%*.*s: ";
+    os_time tv;
+    os_timeSec sec;
+    int32_t nsec;
+
+    (void)os_threadGetThreadName(buf, sizeof(buf));
```

Comment:
Avoids a call to pthread_getspecific. The code will always have to check if a thread's name is set or not because of foreign (threads not created by Cyclone) threads could call the log function.